### PR TITLE
Fix iree prof output for GPU zones.

### PR DESCRIPTION
**Pull request by protobird-git**
1) Thread name retrival is different.
2) Thread ID can be conflicted with ones of CPU zones. Negates it safely
   since thread id is unsigned.
3) Calculate thread duration correctly to cover the case when top-level
   zones are multiple and separated.
4) Added unittests for duration merge. But, iree_cc_test() doesn't work
   in tools/ dir.

### CODE REVIEW

File: tools/iree-prof-output-xplane.cc [hunk 1]
```diff
@@ -11,6 +11,7 @@
 
 #include "build_tools/third_party/tsl/xplane.pb.h"
 #include "third_party/abseil-cpp/absl/container/flat_hash_map.h"
+#include "third_party/abseil-cpp/absl/log/log.h"
```
Comment (okkwon) [inline tools/iree-prof-output-xplane.cc]: nit: seems unused.

File: tools/iree-prof-output-stdout.cc [hunk 2]
```diff
@@ -57,37 +58,30 @@ std::string GetDurationStr(int64_t duration_ns,
...
 
-// Returns the duration per thread, i.e from the earliest start to the latest
-// end of zones running on each thread.
+// Returns the duration per thread, i.e merged durations of all zones running on
+// each thread.
 template <typename T>
 absl::flat_hash_map<int, int64_t> GetThreadDurations(
     const tracy::Worker& worker,
     const tracy::unordered_flat_map<int16_t, T>& zones,
     const std::vector<std::string>& thread_substrs) {
```
Comment (okkwon) [inline tools/iree-prof-output-stdout.cc:67]: I think having substrings as a filter option made the logic more complicated.

We have the following items:

1. Total execution time.
2. Memory transfer & dispatch times per thread

I think the entities from 2) are distinct and they are not supposed to overlap.

Just from the above, we can have
1. accumulated dispatch and memory transfer times for each thread
2. averaged dispatch and memory transfer times time (for all threads)

Can the assumption with no overlapping simply the logic?

File: tools/iree-prof-output-utils.cc [hunk 3]
```diff
@@ -22,8 +23,60 @@ const char* GetZoneName(const tracy::Worker& worker,
...
+  CHECK_LE(least_index_greater_than_or_equal_to_start,
+           greatest_index_less_than_or_equal_to_end + 1);
+
+  bool start_is_within_an_existing_duration =
+      least_index_greater_than_or_equal_to_start % 2 != 0;
+  bool end_is_within_an_existing_duration =
+      greatest_index_less_than_or_equal_to_end % 2 == 0;
+
+  // Merge all durations between start, end into one duration
+  merged.erase(merged.begin() + least_index_greater_than_or_equal_to_start,
```
Comment (okkwon) [inline tools/iree-prof-output-utils.cc]: nit: The function can be expensive to run depending on the input's characteristic. If z.second.zones is already sorted in the time, we might not really touch the items at the end, but if it is not, the erase and insert in the middle of vector can be very expensive.

It reminds me of this coding questions:
https://leetcode.com/problems/merge-intervals/solutions/4519554/step-by-step-explanation-easy-to-understand-in-cpp-javal-js-python/

**Review by okkwon — COMMENTED** [review]
Thanks for the great progress!

I left a few comments and one of the comments is about simplifying the assumption about not having any overlap on the filtered zones. Please let me know how you think. Thanks!

File: tools/iree-prof-output-stdout.cc (hunk 2 above)
Comment (protobird-git) [reply tools/iree-prof-output-stdout.cc:67]: Changed to get durations only from top-level zones which are listed in the timeline in thread data. It's just simple summation, not comparing start/end of zones.

File: tools/iree-prof-output-xplane.cc (hunk 1 above)
Comment (protobird-git) [reply tools/iree-prof-output-xplane.cc]: Removed

File: tools/iree-prof-output-utils.cc (hunk 3 above)
Comment (protobird-git) [reply tools/iree-prof-output-utils.cc]: Changed to use only zone in the timeline of thread data which reduced most of expensive operations.

File: tools/iree-prof-output-utils.cc [hunk 4]
```diff
@@ -8,75 +8,140 @@
...
+  return duration;
+}
 
-  int least_index_greater_than_or_equal_to_start = 0;
-  while (least_index_greater_than_or_equal_to_start < merged.size() &&
-         start > merged[least_index_greater_than_or_equal_to_start]) {
-    ++least_index_greater_than_or_equal_to_start;
+uint64_t DecompressOrFixGpuThreadId(const tracy::Worker& worker,
+                                    uint16_t gpu_thread_id) {
+  if (gpu_thread_id < worker.GetThreadData().size()) {
```
Comment (okkwon) [inline tools/iree-prof-output-utils.cc:59]: Please add a comment why this case and the else case happen.

File: tools/iree-prof-output-utils.cc [hunk 5]
```diff
@@ -8,75 +8,140 @@
...
+  return *event;
 }
 
-void MergeDuration(int64_t start, int64_t end, std::vector<int64_t>& merged) {
-  CHECK_LE(start, end);
-  if (start == end) {
-    return;  // No changes.
+template <typename T>
+int64_t SumDurationInTimeline(const tracy::Worker& worker,
+                              const tracy::Vector<T>& timeline) {
```
Comment (okkwon) [inline tools/iree-prof-output-utils.cc:38]: Please add a comment what `timeline` is.

File: tools/iree-prof-output-utils.cc [hunk 6]
```diff
@@ -8,75 +8,140 @@
...
 
-  if (!start_is_within_an_existing_duration) {
-    merged.insert(merged.begin() + least_index_greater_than_or_equal_to_start,
-                  start);
+template <>
+int64_t GetThreadDuration<tracy::Worker::SourceLocationZones>(
+    const tracy::Worker& worker,
+    int thread_id) {
+  const auto* data = worker.GetThreadData(worker.DecompressThread(thread_id));
+  if (!data->timeline.is_magic()) {
```
Comment (okkwon) [inline tools/iree-prof-output-utils.cc:130]: Please add a comment what `magic` is.

**Review by okkwon — APPROVED** [review]
Looks great. I left a few minor comments. Thanks!

File: tools/iree-prof-output-utils.cc (hunk 5 above)
Comment (protobird-git) [reply tools/iree-prof-output-utils.cc:38]: Done

File: tools/iree-prof-output-utils.cc (hunk 4 above)
Comment (protobird-git) [reply tools/iree-prof-output-utils.cc:59]: Done

File: tools/iree-prof-output-utils.cc (hunk 6 above)
Comment (protobird-git) [reply tools/iree-prof-output-utils.cc:130]: Done

**Pull request merged**